### PR TITLE
Users CRUD and Authorize using Cancan

### PR DESCRIPTION
**Pull request by SubrahmanyamKonduri-BTC**
Admin create/edit/view/destroy other users and authorizing access to various sections of application using cancan.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -5,6 +5,16 @@ class ApplicationController < ActionController::Base
 
   before_action :configure_permitted_parameters, if: :devise_controller?
 
+  rescue_from CanCan::AccessDenied do |exception|
+    redirect_to root_url, :alert => exception.message
```
Comment (amitpatelx) [inline app/controllers/application_controller.rb]: Use Ruby 2 style hash notation

Comment (SubrahmanyamKonduri-BTC) [inline app/controllers/application_controller.rb]: Ok

File: app/controllers/user_controller.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    @users = User.where.not(id: current_user.id)
+  end
+
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
+    @user.admin_id = current_user.id if current_user.role == "Admin"
```
Comment (amitpatelx) [inline app/controllers/user_controller.rb]: `current_user.role == "Admin"` condition checking was repeated at many places even in vi. I don't remember the PR but I would suggest to create a method `is_admin` in ApplicationController. @PragatiDoshi-BTC has created similar method in View helper. 

Co-ordinate with her and extract common method in Application helper and make this method to avail to view helper as well.

File: app/controllers/user_controller.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    if @user.save
+      flash[:notice] = "Successfully created User."
+      redirect_to user_index_path
+    else
+      flash[:error] = @user.errors.full_messages.join(', ')
```
Comment (amitpatelx) [inline app/controllers/user_controller.rb]: `flash[:error] = @user.errors.full_messages.join(', ')` is repeated at many place.  Move this logic to application controller.

File: app/controllers/user_controller.rb (hunk 2 above)
Comment (SubrahmanyamKonduri-BTC) [inline app/controllers/user_controller.rb]: Ok

File: app/controllers/user_controller.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      render action: 'new'
+    end
+  end
+
+  def edit
+    @user = User.find(params[:id])
```
Comment (amitpatelx) [inline app/controllers/user_controller.rb]: `load_and_authorize_resource` is already loading user and setting `@user`, isn't it?

File: app/models/ability.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  def initialize(user)
+    user ||= User.new
+
+    if user.role == "Admin"
+      can :manage, :all
+    else
+      cannot [User, Client]
+    end
+
+    # Define abilities for the passed in user here. For example:
```
Comment (amitpatelx) [inline app/models/ability.rb]: Remove all the commented code

File: app/models/user.rb [hunk 6]
```diff
@@ -5,14 +5,26 @@ class User < ActiveRecord::Base
   devise :database_authenticatable, :registerable, :confirmable,
          :recoverable, :rememberable, :trackable, :validatable
 
+  ## Validations
+  validates_presence_of :role, if: :other_roles
```
Comment (amitpatelx) [inline app/models/user.rb]: Use Rails 4 style validation helpers

File: app/models/user.rb [hunk 7]
```diff
@@ -5,14 +5,26 @@ class User < ActiveRecord::Base
...
 
+  ## Constants
+  ROLE = [["Auditor","auditor"], ["Accountant","accountant"], ["Manager","manager"]]
+
   ## Callbacks
   after_create :assign_role_and_admin_id
 
+  # Override Devise::Confirmable#after_confirmation
+  def after_confirmation
+    send_reset_password_instructions if ((role != "Admin") && (admin_id != id))
```
Comment (amitpatelx) [inline app/models/user.rb]: move `((role != "Admin") && (admin_id != id))` checking with appropriate method name like `is_admin?`

File: app/models/user.rb [hunk 8]
```diff
@@ -20,4 +32,12 @@ def assign_role_and_admin_id
     end
   end
 
+  def other_roles
+    ((admin_id.present?) && (admin_id != id))
+  end
+
+  def check_role
+    errors.add(:role, "is not valid, please select a valid role.") if !(User::ROLE).map{|a| a[1]}.include?(role)
```
Comment (amitpatelx) [inline app/models/user.rb]: use `unless` over `if !`. Also can't we make this check compact?

File: app/views/layouts/application.html.haml [hunk 9]
```diff
@@ -12,15 +12,19 @@
...
-              = link_to "Invoices", ""
-            %li{class: "#{top_navbar_active_class_name('clients','index')}"} 
-              = link_to "Clients", clients_path
+          -if current_user.present?
+            %ul.nav.navbar-nav
+              %li{class: "#{top_navbar_active_class_name('home','index')}"} 
+                = link_to "Home", root_path
+              %li{class: "#{top_navbar_active_class_name('invoices','index')}"} 
+                = link_to "Invoices", ""
+              - if current_user.role == "Admin"
```
Comment (amitpatelx) [inline app/views/layouts/application.html.haml]: Use helper method `is_admin?`

File: app/models/user.rb [hunk 10]
```diff
@@ -5,14 +5,26 @@ class User < ActiveRecord::Base
...
+  validates_presence_of :role, if: :other_roles
+  validate :check_role, if: :other_roles
+
   ## Associations
   belongs_to :company
   has_many :invoices, through: :company
   has_many :clients, through: :company
 
+  ## Constants
+  ROLE = [["Auditor","auditor"], ["Accountant","accountant"], ["Manager","manager"]]
```
Comment (amitpatelx) [inline app/models/user.rb]: `ROLE` should be `ROLES`, an array of roles in downcase. To use roles as option to select box create constant `ROLES_OPTIONS` in respective helper module and use User::ROLES to give values to ROLES_OPTIONS`

File: app/views/user/index.html.haml [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+        \.
+      %td= user.email
+      %td= user.first_name
+      %td= user.last_name
+      %td= user.role
+      %td
+        %div
+          .inline.position-relative
+            = link_to "View", user_path(user)
+            &nbsp;&nbsp;
```
Comment (amitpatelx) [inline app/views/user/index.html.haml]: Use bootstrap classes instead for `&nbsp;`

File: app/views/user/new.html.haml [hunk 12]
```diff
@@ -0,0 +1,30 @@
+%h1 New User
+%br/
```
Comment (amitpatelx) [inline app/views/user/new.html.haml]: Don't use `br` anywhere. User appropriate classes. Study bootstrap css framework

File: app/views/user/new.html.haml [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+= form_for @user, url: '/user', html: { method: :post } do |f|
```
Comment (amitpatelx) [inline app/views/user/new.html.haml]: Check the layout of the form of the theme
http://responsiweb.com/themes/preview/ace/1.3.3/form-elements.html

Your form has to follows standard of bootstrap

File: app/views/user/new.html.haml [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+    = f.label :last_name
+    %br/
+    = f.text_field :last_name
+  .field
+    = f.label "Role"
+    %br/
+    = f.select :role, options_for_select(User::ROLE, f.object.role)
+  .actions
+    %br/
+    = f.submit "Create"
```
Comment (amitpatelx) [inline app/views/user/new.html.haml]: When user press the submit button, it should get disabled so user cannot press again intentionally or accidently.

Check `disable_with` option on http://api.rubyonrails.org/classes/ActionView/Helpers/FormTagHelper.html

File: app/views/user/new.html.haml [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+  .field
+    = f.label :first_name
+    %br/
+    = f.text_field :first_name
+  .field
+    = f.label :last_name
+    %br/
+    = f.text_field :last_name
+  .field
+    = f.label "Role"
```
Comment (amitpatelx) [inline app/views/user/new.html.haml]: Use Symbol instead of String for label

File: app/views/user/show.html.haml [hunk 16]
```diff
@@ -0,0 +1,18 @@
+%h1 User Profile:
+%br/
+.profile-user-info.profile-user-info-striped
+  .profile-info-row
+    %b Email: &nbsp;
```
Comment (amitpatelx) [inline app/views/user/show.html.haml]: Use bootstrap class instead `&nbsp;`. Check this across all the pages

**Pull request merged**